io.home-assistant/index.js

```diff
@@ -2,12 +2,266 @@
 //
 // This file is part of io.home-assistant
 //
-// Copyright 2019 Giovanni Campagna <gcampagn@cs.stanford.edu>,Silei Xu <silei@cs.stanford.edu>
+// Copyright 2019 Giovanni Campagna <gcampagn@cs.stanford.edu>
 //
 // See LICENSE for details
 "use strict";
 
 const Tp = require('thingpedia');
+const WebSocket = require('ws');
+const HomeAssistant = require('home-assistant-js-websocket');
+
+// FIXME make configurable
+const HASS_URL = 'http://hassio.local:8123';
+
+class HomeAssistantDevice extends Tp.BaseDevice {
+    constructor(engine, state, master, entityId) {
+        super(engine, state);
+
+        this.master = master;
+        this._entityId = entityId;
+        this.uniqueId = master.uniqueId + '/' + entityId;
+        this.name = this.state.friendly_name;
+    }
+
+    updateState(state) {
+        super.updateState(state);
+        this.name = this.state.friendly_name;
+    }
+
+    _callService(domain, service, data = {}) {
+        data.entity_id = this._entityId;
+        return HomeAssistant.callService(this.master.connection, domain, service, data);
+    }
+}
+
+class HomeAssistantLightbulbDevice extends HomeAssistantDevice {
+    constructor(engine, state, master, entityId) {
+        super(engine, state, master, entityId);
+
+        this.uniqueId = master.uniqueId + '/' + entityId;
+    }
+
+    async do_set_power({ power }) {
+        if (power === 'on')
+            await this._callService('light', 'turn_on');
+        else
+            await this._callService('light', 'turn_off');
+    }
+
+    async do_alert_long() {
+        await this._callService('light', 'turn_on', { flash: 'long' });
+    }
+
+    async do_color_loop() {
+        await this._callService('light', 'turn_on', { effect: 'colorloop' });
+    }
+}
+
+class HomeAssistantDeviceSet extends Tp.Helpers.ObjectSet.Base {
```

Comment:
What is the difference between `ObjectSet`, `SimpleObjectSet`, and `DeviceView` in `thingengine-core`?
(I never touched this part of the code, I figure it's good to learn now)

Reply:
ObjectSet is generic a way to expose a dynamic set of objects backed by some underlying store.
ObjectSet.Base is an interface, which can have different implementations. It is a read only interface: objects come and go from the set according to the implementation, and the interface user listens to object-added / object-removed events.

ObjectSet.Simple is an implementation of the ObjectSet.Base interface backed by a JS Map, that does nothing on its own and is driven by API calls.

DeviceView is another implementation of ObjectSet. It wraps an ObjectSet of Thingpedia devices (usually, the whole DeviceDatabase), it flattens any nested subdevice collection, and filters according to a ThingTalk selector. In other words, it is the dynamic set of devices that match a ThingTalk selector.

(ThingTalk selectors used to allow choosing devices by arbitrary labels, and we used to care about long-lived rules a lot more. Nowadays, we choose devices by unique ID, and we mostly care about one-off commands where dynamic changes are not important. Yet, the code is still there for the future.)